File: usr/share/ashypass/ui/generator_view.py
#!/usr/bin/env python3
"""Ashy Pass - Generator View"""

import gi
gi.require_version('Gtk', '4.0')
gi.require_version('Adw', '1')

from gi.repository import Gtk, Adw, GLib
from core.generator import PasswordGenerator, PasswordConfig
from utils.clipboard import ClipboardManager
from utils.i18n import _
from core.config import *
from core.config import load_settings, save_settings


class GeneratorView(Adw.NavigationPage):
    """Password generator view"""
    
    def __init__(self):
        super().__init__(title=_("Generator"))
        self.generator = PasswordGenerator()
        self.clipboard = ClipboardManager()
        self.current_password = ""
        self.settings = load_settings()
        self._build_ui()
        self._load_preferences()
        self._generate_password()
    
    def _build_ui(self) -> None:
        """Build the user interface"""
        scrolled = Gtk.ScrolledWindow()
        scrolled.set_policy(Gtk.PolicyType.NEVER, Gtk.PolicyType.AUTOMATIC)

        content = Gtk.Box(orientation=Gtk.Orientation.VERTICAL)
        content.set_margin_top(12)
        content.set_margin_bottom(12)
        content.set_margin_start(12)
        content.set_margin_end(12)
        content.set_spacing(24)
        
        # Password display
        group = Adw.PreferencesGroup()
        group.set_title(_("Generated Password"))

        # Password row - same style as Strength row
        password_row = Adw.ActionRow()
        password_row.set_title(_("Password"))

        # Password label - selectable, monospace, left-aligned in a scrolled window
        password_scroll = Gtk.ScrolledWindow()
        password_scroll.set_policy(Gtk.PolicyType.AUTOMATIC, Gtk.PolicyType.NEVER)
        password_scroll.set_max_content_width(400)
        password_scroll.set_propagate_natural_width(True)

        self.password_label = Gtk.Label()
        self.password_label.set_selectable(True)
        self.password_label.add_css_class("monospace")
        self.password_label.add_css_class("title-3")
        self.password_label.set_xalign(0.0)  # Left align

        password_scroll.set_child(self.password_label)
        password_row.add_suffix(password_scroll)

        group.add(password_row)
        
        # Strength indicator
        strength_row = Adw.ActionRow()
        strength_row.set_title(_("Strength"))
        self.strength_label = Gtk.Label()
        self.strength_label.add_css_class("title-4")
        strength_row.add_suffix(self.strength_label)
        group.add(strength_row)
        
        self.strength_bar = Gtk.LevelBar()
        self.strength_bar.set_mode(Gtk.LevelBarMode.CONTINUOUS)
        self.strength_bar.set_min_value(0)
        self.strength_bar.set_max_value(100)
        self.strength_bar.set_margin_top(6)
        self.strength_bar.set_margin_bottom(6)
        self.strength_bar.set_margin_start(12)
        self.strength_bar.set_margin_end(12)
        group.add(self.strength_bar)
        
        # Buttons
        btn_box = Gtk.Box(orientation=Gtk.Orientation.HORIZONTAL)
        btn_box.set_spacing(12)
        btn_box.set_halign(Gtk.Align.CENTER)
        btn_box.set_margin_top(12)
        
        copy_btn = Gtk.Button(label=_("Copy to Clipboard"))
        copy_btn.add_css_class("pill")
        copy_btn.add_css_class("suggested-action")
        copy_btn.connect("clicked", self._on_copy_clicked)
        btn_box.append(copy_btn)

        regen_btn = Gtk.Button(label=_("Generate New"))
        regen_btn.add_css_class("pill")
        regen_btn.connect("clicked", lambda _: self._generate_password())
        btn_box.append(regen_btn)
        
        group.add(btn_box)
        content.append(group)
        
        # Type selector
        type_group = Adw.PreferencesGroup()
        type_group.set_title(_("Generation Type"))

        type_box = Gtk.Box(orientation=Gtk.Orientation.HORIZONTAL)
        type_box.set_halign(Gtk.Align.CENTER)
        type_box.add_css_class("linked")

        password_btn = Gtk.ToggleButton(label=_("Password"))
        password_btn.set_active(True)
        password_btn.connect("toggled", self._on_type_changed, "password")
        type_box.append(password_btn)

        passphrase_btn = Gtk.ToggleButton(label=_("Passphrase"))
        passphrase_btn.set_group(password_btn)
        passphrase_btn.connect("toggled", self._on_type_changed, "passphrase")
        type_box.append(passphrase_btn)

        pin_btn = Gtk.ToggleButton(label=_("PIN"))
        pin_btn.set_group(password_btn)
        pin_btn.connect("toggled", self._on_type_changed, "pin")
        type_box.append(pin_btn)
        
        type_group.add(type_box)
        content.append(type_group)
        
        # Options stack
        self.options_stack = Gtk.Stack()
        self.options_stack.set_transition_type(Gtk.StackTransitionType.SLIDE_LEFT_RIGHT)
        self.options_stack.add_titled(self._create_password_options(), "password", "Password")
        self.options_stack.add_titled(self._create_passphrase_options(), "passphrase", "Passphrase")
        self.options_stack.add_titled(self._create_pin_options(), "pin", "PIN")
        content.append(self.options_stack)

        scrolled.set_child(content)
        self.set_child(scrolled)
    
    def _create_password_options(self) -> Adw.PreferencesGroup:
        """Create password generation options"""
        group = Adw.PreferencesGroup()
        group.set_title(_("Password Options"))

        self.length_spin = Adw.SpinRow()
        self.length_spin.set_title(_("Length"))
        self.length_spin.set_adjustment(Gtk.Adjustment(value=DEFAULT_PASSWORD_LENGTH,
            lower=MIN_PASSWORD_LENGTH, upper=MAX_PASSWORD_LENGTH, step_increment=1))
        self.length_spin.connect("changed", lambda _: self._on_option_changed())
        group.add(self.length_spin)

        self.uppercase_switch = Adw.SwitchRow(title=_("Uppercase Letters (A-Z)"), active=True)
        self.uppercase_switch.connect("notify::active", lambda *_: self._on_option_changed())
        group.add(self.uppercase_switch)

        self.lowercase_switch = Adw.SwitchRow(title=_("Lowercase Letters (a-z)"), active=True)
        self.lowercase_switch.connect("notify::active", lambda *_: self._on_option_changed())
        group.add(self.lowercase_switch)

        self.digits_switch = Adw.SwitchRow(title=_("Digits (0-9)"), active=True)
        self.digits_switch.connect("notify::active", lambda *_: self._on_option_changed())
        group.add(self.digits_switch)

        self.symbols_switch = Adw.SwitchRow(title=_("Symbols (!@#$...)"), active=True)
        self.symbols_switch.connect("notify::active", lambda *_: self._on_option_changed())
        group.add(self.symbols_switch)

        self.ambiguous_switch = Adw.SwitchRow(title=_("Exclude Ambiguous Characters"), active=True)
        self.ambiguous_switch.set_subtitle(_("Avoid characters like 0, O, 1, l, I"))
        self.ambiguous_switch.connect("notify::active", lambda *_: self._on_option_changed())
        group.add(self.ambiguous_switch)
        
        return group
    
    def _create_passphrase_options(self) -> Adw.PreferencesGroup:
        """Create passphrase generation options"""
        group = Adw.PreferencesGroup()
        group.set_title(_("Passphrase Options"))

        self.words_spin = Adw.SpinRow(title=_("Number of Words"))
        self.words_spin.set_adjustment(Gtk.Adjustment(value=DEFAULT_PASSPHRASE_WORDS,
            lower=MIN_PASSPHRASE_WORDS, upper=MAX_PASSPHRASE_WORDS, step_increment=1))
        self.words_spin.connect("changed", lambda _: self._on_option_changed())
        group.add(self.words_spin)

        self.separator_entry = Adw.EntryRow(title=_("Separator"), text="-")
        self.separator_entry.connect("changed", lambda _: self._on_option_changed())
        group.add(self.separator_entry)

        self.capitalize_switch = Adw.SwitchRow(title=_("Capitalize Words"), active=True)
        self.capitalize_switch.connect("notify::active", lambda *_: self._on_option_changed())
        group.add(self.capitalize_switch)

        self.add_number_switch = Adw.SwitchRow(title=_("Add Number at End"), active=True)
        self.add_number_switch.connect("notify::active", lambda *_: self._on_option_changed())
        group.add(self.add_number_switch)
        
        return group
    
    def _create_pin_options(self) -> Adw.PreferencesGroup:
        """Create PIN generation options"""
        group = Adw.PreferencesGroup()
        group.set_title(_("PIN Options"))

        self.pin_length_spin = Adw.SpinRow(title=_("Length"))
        self.pin_length_spin.set_adjustment(Gtk.Adjustment(value=DEFAULT_PIN_LENGTH,
            lower=MIN_PIN_LENGTH, upper=MAX_PIN_LENGTH, step_increment=1))
        self.pin_length_spin.connect("changed", lambda _: self._on_option_changed())
        group.add(self.pin_length_spin)
        
        return group
    
    def _on_type_changed(self, button: Gtk.ToggleButton, type_name: str) -> None:
        """Handle generation type change"""
        if button.get_active():
            self.options_stack.set_visible_child_name(type_name)
            self._generate_password()
    
    def _generate_password(self) -> None:
        """Generate password based on current settings"""
        current_type = self.options_stack.get_visible_child_name()
        
        try:
            if current_type == "password":
                config = PasswordConfig(
                    length=int(self.length_spin.get_value()),
                    use_uppercase=self.uppercase_switch.get_active(),
                    use_lowercase=self.lowercase_switch.get_active(),
                    use_digits=self.digits_switch.get_active(),
                    use_symbols=self.symbols_switch.get_active(),
                    exclude_ambiguous=self.ambiguous_switch.get_active(),
                )
                self.current_password = self.generator.generate_password(config)
            elif current_type == "passphrase":
                self.current_password = self.generator.generate_passphrase(
                    num_words=int(self.words_spin.get_value()),
                    separator=self.separator_entry.get_text(),
                    capitalize=self.capitalize_switch.get_active(),
                    add_number=self.add_number_switch.get_active(),
                )
            elif current_type == "pin":
                self.current_password = self.generator.generate_pin(int(self.pin_length_spin.get_value()))
            
            self.password_label.set_text(self.current_password)
            self._update_strength_indicator()
        except ValueError as e:
            print(f"Error: {e}")
    
    def _update_strength_indicator(self) -> None:
        """Update password strength indicator"""
        score, level = self.generator.check_password_strength(self.current_password)
        self.strength_label.set_text(_(level))
        self.strength_bar.set_value(score)
        
        # Update color
        self.strength_label.remove_css_class("success")
        self.strength_label.remove_css_class("warning")
        self.strength_label.remove_css_class("error")
        
        if score >= 80:
            self.strength_label.add_css_class("success")
        elif score >= 40:
            self.strength_label.add_css_class("warning")
        else:
            self.strength_label.add_css_class("error")

    def _load_preferences(self) -> None:
        """Load saved generator preferences"""
        gen_prefs = self.settings.get("generator", {})

        # Load password options
        if "length" in gen_prefs:
            self.length_spin.set_value(gen_prefs["length"])
        if "uppercase" in gen_prefs:
            self.uppercase_switch.set_active(gen_prefs["uppercase"])
        if "lowercase" in gen_prefs:
            self.lowercase_switch.set_active(gen_prefs["lowercase"])
        if "digits" in gen_prefs:
            self.digits_switch.set_active(gen_prefs["digits"])
        if "symbols" in gen_prefs:
            self.symbols_switch.set_active(gen_prefs["symbols"])
        if "exclude_ambiguous" in gen_prefs:
            self.ambiguous_switch.set_active(gen_prefs["exclude_ambiguous"])

        # Load passphrase options
        if "passphrase_words" in gen_prefs:
            self.words_spin.set_value(gen_prefs["passphrase_words"])
        if "passphrase_separator" in gen_prefs:
            self.separator_entry.set_text(gen_prefs["passphrase_separator"])
        if "passphrase_capitalize" in gen_prefs:
            self.capitalize_switch.set_active(gen_prefs["passphrase_capitalize"])
        if "passphrase_add_number" in gen_prefs:
            self.add_number_switch.set_active(gen_prefs["passphrase_add_number"])

        # Load PIN options
        if "pin_length" in gen_prefs:
            self.pin_length_spin.set_value(gen_prefs["pin_length"])

    def _save_preferences(self) -> None:
        """Save current generator preferences"""
        gen_prefs = {
            "length": int(self.length_spin.get_value()),
            "uppercase": self.uppercase_switch.get_active(),
            "lowercase": self.lowercase_switch.get_active(),
            "digits": self.digits_switch.get_active(),
            "symbols": self.symbols_switch.get_active(),
            "exclude_ambiguous": self.ambiguous_switch.get_active(),
            "passphrase_words": int(self.words_spin.get_value()),
            "passphrase_separator": self.separator_entry.get_text(),
            "passphrase_capitalize": self.capitalize_switch.get_active(),
            "passphrase_add_number": self.add_number_switch.get_active(),
            "pin_length": int(self.pin_length_spin.get_value()),
        }

        self.settings["generator"] = gen_prefs
        save_settings(self.settings)

    def _on_option_changed(self) -> None:
        """Handle option change - save preferences and regenerate"""
        self._save_preferences()
        self._generate_password()

    def _on_copy_clicked(self, button: Gtk.Button) -> None:
        """Handle copy button click"""
        if self.current_password:
            self.clipboard.copy_text(self.current_password)
            self.activate_action("app.show-toast", GLib.Variant.new_string(_("Password copied to clipboard")))

File: crates/ashypass-app/src/ui/generator_view.rs
//! Password generator view — task #8.
//!
//! Top: generated password display + strength bar + Copy/Regenerate buttons.
//! Middle: type selector (Password / Passphrase / PIN) toggling an options Stack.
//! Bottom: per-type options (length, character classes, words, separator, etc.).

use crate::tr;
use adw::prelude::*;
use ashypass_core::config::{
    DEFAULT_PASSPHRASE_WORDS, DEFAULT_PASSWORD_LENGTH, DEFAULT_PIN_LENGTH, MAX_PASSPHRASE_WORDS,
    MAX_PASSWORD_LENGTH, MAX_PIN_LENGTH, MIN_PASSPHRASE_WORDS, MIN_PASSWORD_LENGTH, MIN_PIN_LENGTH,
};
use ashypass_core::generator::{
    generate_passphrase, generate_password, generate_pin, PasswordConfig,
};
use ashypass_core::strength::legacy_score as check_password_strength;
use gtk::glib;
use std::cell::RefCell;
use std::rc::Rc;

pub struct GeneratorView {
    pub root: gtk::Box,
    #[expect(dead_code, reason = "keeps view model alive for signal handlers")]
    inner: Rc<Inner>,
}

struct Inner {
    toast: adw::ToastOverlay,
    current_password: RefCell<String>,

    password_label: gtk::Label,
    strength_label: gtk::Label,
    strength_bar: gtk::LevelBar,

    options_stack: gtk::Stack,

    // password options
    length_spin: adw::SpinRow,
    uppercase_switch: adw::SwitchRow,
    lowercase_switch: adw::SwitchRow,
    digits_switch: adw::SwitchRow,
    symbols_switch: adw::SwitchRow,
    ambiguous_switch: adw::SwitchRow,

    // passphrase options
    words_spin: adw::SpinRow,
    separator_entry: adw::EntryRow,
    capitalize_switch: adw::SwitchRow,
    add_number_switch: adw::SwitchRow,

    // pin options
    pin_length_spin: adw::SpinRow,
}

impl GeneratorView {
    pub fn new(toast: adw::ToastOverlay) -> Self {
        let root = gtk::Box::new(gtk::Orientation::Vertical, 0);
        root.set_vexpand(true);
        root.set_hexpand(true);

        let scrolled = gtk::ScrolledWindow::builder()
            .hscrollbar_policy(gtk::PolicyType::Never)
            .vscrollbar_policy(gtk::PolicyType::Automatic)
            .vexpand(true)
            .hexpand(true)
            .build();
        let content = gtk::Box::builder()
            .orientation(gtk::Orientation::Vertical)
            .spacing(24)
            .margin_top(12)
            .margin_bottom(12)
            .margin_start(12)
            .margin_end(12)
            .build();

        // ---- Generated password group ----
        let pwd_group = adw::PreferencesGroup::builder()
            .title(tr!("Generated Password"))
            .build();

        let password_row = adw::ActionRow::builder().title(tr!("Password")).build();
        let password_scroll = gtk::ScrolledWindow::builder()
            .hscrollbar_policy(gtk::PolicyType::Automatic)
            .vscrollbar_policy(gtk::PolicyType::Never)
            .max_content_width(400)
            .propagate_natural_width(true)
            .build();
        let password_label = gtk::Label::new(None);
        password_label.set_selectable(true);
        password_label.add_css_class("monospace");
        password_label.add_css_class("title-3");
        password_label.set_xalign(0.0);
        password_scroll.set_child(Some(&password_label));
        password_row.add_suffix(&password_scroll);
        pwd_group.add(&password_row);

        let strength_row = adw::ActionRow::builder().title(tr!("Strength")).build();
        let strength_label = gtk::Label::new(None);
        strength_label.add_css_class("title-4");
        strength_row.add_suffix(&strength_label);
        pwd_group.add(&strength_row);

        content.append(&pwd_group);

        let strength_bar = gtk::LevelBar::builder()
            .mode(gtk::LevelBarMode::Continuous)
            .min_value(0.0)
            .max_value(100.0)
            .margin_top(6)
            .margin_bottom(6)
            .margin_start(12)
            .margin_end(12)
            .build();
        content.append(&strength_bar);

        let btn_box = gtk::Box::builder()
            .orientation(gtk::Orientation::Horizontal)
            .spacing(12)
            .halign(gtk::Align::Center)
            .margin_top(6)
            .build();
        let copy_btn = gtk::Button::with_label(tr!("Copy to Clipboard"));
        copy_btn.add_css_class("pill");
        copy_btn.add_css_class("suggested-action");
        btn_box.append(&copy_btn);
        let regen_btn = gtk::Button::with_label(tr!("Generate New"));
        regen_btn.add_css_class("pill");
        btn_box.append(&regen_btn);
        content.append(&btn_box);

        // ---- Type selector ----
        let type_group = adw::PreferencesGroup::builder()
            .title(tr!("Generation Type"))
            .build();
        let type_box = gtk::Box::builder()
            .orientation(gtk::Orientation::Horizontal)
            .halign(gtk::Align::Center)
            .build();
        type_box.add_css_class("linked");

        let pwd_tg = gtk::ToggleButton::with_label(tr!("Password"));
        pwd_tg.set_active(true);
        type_box.append(&pwd_tg);
        let pass_tg = gtk::ToggleButton::with_label(tr!("Passphrase"));
        pass_tg.set_group(Some(&pwd_tg));
        type_box.append(&pass_tg);
        let pin_tg = gtk::ToggleButton::with_label(tr!("PIN"));
        pin_tg.set_group(Some(&pwd_tg));
        type_box.append(&pin_tg);
        type_group.add(&type_box);
        content.append(&type_group);

        // ---- Options stack ----
        let options_stack = gtk::Stack::builder()
            .transition_type(gtk::StackTransitionType::SlideLeftRight)
            .build();

        let (
            pwd_options,
            length_spin,
            uppercase_switch,
            lowercase_switch,
            digits_switch,
            symbols_switch,
            ambiguous_switch,
        ) = build_password_options();
        options_stack.add_named(&pwd_options, Some("password"));

        let (pass_options, words_spin, separator_entry, capitalize_switch, add_number_switch) =
            build_passphrase_options();
        options_stack.add_named(&pass_options, Some("passphrase"));

        let (pin_options, pin_length_spin) = build_pin_options();
        options_stack.add_named(&pin_options, Some("pin"));

        content.append(&options_stack);

        scrolled.set_child(Some(&content));
        root.append(&scrolled);

        let inner = Rc::new(Inner {
            toast,
            current_password: RefCell::new(String::new()),
            password_label,
            strength_label,
            strength_bar,
            options_stack: options_stack.clone(),
            length_spin,
            uppercase_switch,
            lowercase_switch,
            digits_switch,
            symbols_switch,
            ambiguous_switch,
            words_spin,
            separator_entry,
            capitalize_switch,
            add_number_switch,
            pin_length_spin,
        });

        // Wire up callbacks
        {
            let inner_cl = inner.clone();
            copy_btn.connect_clicked(move |_| inner_cl.copy_clicked());
        }
        {
            let inner_cl = inner.clone();
            regen_btn.connect_clicked(move |_| inner_cl.generate());
        }
        {
            let inner_cl = inner.clone();
            pwd_tg.connect_toggled(move |b| {
                if b.is_active() {
                    inner_cl.options_stack.set_visible_child_name("password");
                    inner_cl.generate();
                }
            });
        }
        {
            let inner_cl = inner.clone();
            pass_tg.connect_toggled(move |b| {
                if b.is_active() {
                    inner_cl.options_stack.set_visible_child_name("passphrase");
                    inner_cl.generate();
                }
            });
        }
        {
            let inner_cl = inner.clone();
            pin_tg.connect_toggled(move |b| {
                if b.is_active() {
                    inner_cl.options_stack.set_visible_child_name("pin");
                    inner_cl.generate();
                }
            });
        }

        wire_option_changes(&inner);

        inner.generate();

        Self { root, inner }
    }
}

fn wire_option_changes(inner: &Rc<Inner>) {
    let connect_spin = |s: &adw::SpinRow, inner: &Rc<Inner>| {
        let inner_cl = inner.clone();
        s.connect_changed(move |_| inner_cl.generate());
    };
    let connect_switch = |s: &adw::SwitchRow, inner: &Rc<Inner>| {
        let inner_cl = inner.clone();
        s.connect_active_notify(move |_| inner_cl.generate());
    };
    let connect_entry = |s: &adw::EntryRow, inner: &Rc<Inner>| {
        let inner_cl = inner.clone();
        s.connect_changed(move |_| inner_cl.generate());
    };

    connect_spin(&inner.length_spin, inner);
    connect_switch(&inner.uppercase_switch, inner);
    connect_switch(&inner.lowercase_switch, inner);
    connect_switch(&inner.digits_switch, inner);
    connect_switch(&inner.symbols_switch, inner);
    connect_switch(&inner.ambiguous_switch, inner);

    connect_spin(&inner.words_spin, inner);
    connect_entry(&inner.separator_entry, inner);
    connect_switch(&inner.capitalize_switch, inner);
    connect_switch(&inner.add_number_switch, inner);

    connect_spin(&inner.pin_length_spin, inner);
}

impl Inner {
    fn current_type(&self) -> String {
        self.options_stack
            .visible_child_name()
            .map(|s| s.to_string())
            .unwrap_or_else(|| "password".to_string())
    }

    fn generate(&self) {
        let pwd = match self.current_type().as_str() {
            "password" => {
                let cfg = PasswordConfig {
                    length: self.length_spin.value() as usize,
                    use_uppercase: self.uppercase_switch.is_active(),
                    use_lowercase: self.lowercase_switch.is_active(),
                    use_digits: self.digits_switch.is_active(),
                    use_symbols: self.symbols_switch.is_active(),
                    exclude_ambiguous: self.ambiguous_switch.is_active(),
                    custom_symbols: String::new(),
                };
                match generate_password(&cfg) {
                    Ok(p) => p,
                    Err(e) => {
                        self.password_label
                            .set_text(&format!("{}: {e}", tr!("Error")));
                        return;
                    }
                }
            }
            "passphrase" => generate_passphrase(
                self.words_spin.value() as usize,
                self.separator_entry.text().as_str(),
                self.capitalize_switch.is_active(),
                self.add_number_switch.is_active(),
            ),
            "pin" => generate_pin(self.pin_length_spin.value() as usize),
            _ => return,
        };

        self.password_label.set_text(&pwd);
        self.update_strength(&pwd);
        *self.current_password.borrow_mut() = pwd;
    }

    fn update_strength(&self, pwd: &str) {
        let (score, level) = check_password_strength(pwd);
        self.strength_label
            .set_text(crate::ui::i18n::localized_strength_label(level));
        self.strength_bar.set_value(score as f64);

        self.strength_label.remove_css_class("success");
        self.strength_label.remove_css_class("warning");
        self.strength_label.remove_css_class("error");
        if score >= 80 {
            self.strength_label.add_css_class("success");
        } else if score >= 40 {
            self.strength_label.add_css_class("warning");
        } else {
            self.strength_label.add_css_class("error");
        }
    }

    fn copy_clicked(&self) {
        let pwd = self.current_password.borrow().clone();
        if pwd.is_empty() {
            return;
        }
        let seconds = ashypass_core::settings::Settings::load().clipboard_clear;
        crate::clipboard::copy(&pwd, seconds);
        let toast = adw::Toast::builder()
            .title(tr!("Password copied to clipboard"))
            .timeout(3)
            .build();
        self.toast.add_toast(toast);
    }
}

#[allow(clippy::type_complexity)]
fn build_password_options() -> (
    adw::PreferencesGroup,
    adw::SpinRow,
    adw::SwitchRow,
    adw::SwitchRow,
    adw::SwitchRow,
    adw::SwitchRow,
    adw::SwitchRow,
) {
    let group = adw::PreferencesGroup::builder()
        .title(tr!("Password Options"))
        .build();

    let length_adj = gtk::Adjustment::new(
        DEFAULT_PASSWORD_LENGTH as f64,
        MIN_PASSWORD_LENGTH as f64,
        MAX_PASSWORD_LENGTH as f64,
        1.0,
        1.0,
        0.0,
    );
    let length_spin = adw::SpinRow::builder()
        .title(tr!("Length"))
        .adjustment(&length_adj)
        .build();
    group.add(&length_spin);

    let uppercase_switch = adw::SwitchRow::builder()
        .title(tr!("Uppercase Letters (A-Z)"))
        .active(true)
        .build();
    group.add(&uppercase_switch);

    let lowercase_switch = adw::SwitchRow::builder()
        .title(tr!("Lowercase Letters (a-z)"))
        .active(true)
        .build();
    group.add(&lowercase_switch);

    let digits_switch = adw::SwitchRow::builder()
        .title(tr!("Digits (0-9)"))
        .active(true)
        .build();
    group.add(&digits_switch);

    let symbols_switch = adw::SwitchRow::builder()
        .title(tr!("Symbols (!@#$…)"))
        .active(true)
        .build();
    group.add(&symbols_switch);

    let ambiguous_switch = adw::SwitchRow::builder()
        .title(tr!("Exclude Ambiguous Characters"))
        .subtitle(tr!("Avoid characters like 0, O, 1, l, I"))
        .active(true)
        .build();
    group.add(&ambiguous_switch);

    (
        group,
        length_spin,
        uppercase_switch,
        lowercase_switch,
        digits_switch,
        symbols_switch,
        ambiguous_switch,
    )
}

fn build_passphrase_options() -> (
    adw::PreferencesGroup,
    adw::SpinRow,
    adw::EntryRow,
    adw::SwitchRow,
    adw::SwitchRow,
) {
    let group = adw::PreferencesGroup::builder()
        .title(tr!("Passphrase Options"))
        .build();

    let words_adj = gtk::Adjustment::new(
        DEFAULT_PASSPHRASE_WORDS as f64,
        MIN_PASSPHRASE_WORDS as f64,
        MAX_PASSPHRASE_WORDS as f64,
        1.0,
        1.0,
        0.0,
    );
    let words_spin = adw::SpinRow::builder()
        .title(tr!("Number of Words"))
        .adjustment(&words_adj)
        .build();
    group.add(&words_spin);

    let separator_entry = adw::EntryRow::builder()
        .title(tr!("Separator"))
        .text("-")
        .build();
    group.add(&separator_entry);

    let capitalize_switch = adw::SwitchRow::builder()
        .title(tr!("Capitalize Words"))
        .active(true)
        .build();
    group.add(&capitalize_switch);

    let add_number_switch = adw::SwitchRow::builder()
        .title(tr!("Add Number at End"))
        .active(true)
        .build();
    group.add(&add_number_switch);

    (
        group,
        words_spin,
        separator_entry,
        capitalize_switch,
        add_number_switch,
    )
}

fn build_pin_options() -> (adw::PreferencesGroup, adw::SpinRow) {
    let group = adw::PreferencesGroup::builder()
        .title(tr!("PIN Options"))
        .build();

    let pin_adj = gtk::Adjustment::new(
        DEFAULT_PIN_LENGTH as f64,
        MIN_PIN_LENGTH as f64,
        MAX_PIN_LENGTH as f64,
        1.0,
        1.0,
        0.0,
    );
    let pin_length_spin = adw::SpinRow::builder()
        .title(tr!("Length"))
        .adjustment(&pin_adj)
        .build();
    group.add(&pin_length_spin);

    (group, pin_length_spin)
}

#[allow(dead_code)]
fn _unused(_: glib::ControlFlow) {}
